drivers/bus/virtual_bus.c

```diff
@@ -95,6 +103,7 @@ struct bus_type virtual_bus_type = {
 	.suspend = virtbus_suspend,
 	.resume = virtbus_resume,
 	.uevent = virtbus_uevent,
+	.pm = &virtbus_dev_pm_ops,
```

Comment:
can you clarify the commit message? What if the client does want a specific power management support, beyond the generic case? I mean, you could e.g. want a different pm_runtime_autosuspend delay depending on usage - fast for IPC, slow for audio cards that deal with external components.

Reply:
@plbossart the pm ops are not used for setting the autosuspend delay. The generic PM ops are mainly for avoiding the addition of dummy callbacks that do nothing but "return 0" to prevent pm_runtime_get_sync() failing. In case the pm callbacks are set, they will get invoked.

Reply:
I've expanded the commit message a bit. Let me know if this makes more sense.

Reply:
> @plbossart the pm ops are not used for setting the autosuspend delay. The generic PM ops are mainly for avoiding the addition of dummy callbacks that do nothing but "return 0" to prevent pm_runtime_get_sync() failing. In case the pm callbacks are set, they will get invoked.

Can you explain how the pm callbacks can be set, by assigning the .pm field in the .driver substructure?

Reply:
@plbossart Yes, exactly. The generic methods check if the .pm is set and if not return 0.
https://github.com/thesofproject/linux/blob/f9b87470fd5a4f387873665bede8795d97f7358b/drivers/base/power/generic_ops.c#L39

Reply:
ok, that confirms my finding that a .driver structure is the only way to support power management. Thanks for the pointer